File: artifacts/api-server/src/routes/tickets.ts
import { Router, type IRouter } from "express";
import { and, eq, sql } from "drizzle-orm";
import { CreateTicketBody, UpdateTicketBody } from "@workspace/api-zod";
import { db, ticketsTable } from "@workspace/db";
import { requireAuth } from "../middlewares/auth";

const router: IRouter = Router();

function serialize(t: typeof ticketsTable.$inferSelect) {
  return {
    id: t.id,
    subject: t.subject,
    description: t.description,
    priority: t.priority,
    status: t.status,
    externalId: t.externalId,
    createdBy: t.userId,
    relatedMessageId: t.relatedMessageId,
    lastUpdate: t.lastUpdate,
    createdAt: t.createdAt.toISOString(),
    updatedAt: t.updatedAt.toISOString(),
  };
}

router.get("/tickets", requireAuth, async (req, res) => {
  const isAdmin = req.user!.role === "admin";
  const rows = isAdmin
    ? await db.select().from(ticketsTable).orderBy(sql`${ticketsTable.createdAt} DESC`)
    : await db
        .select()
        .from(ticketsTable)
        .where(eq(ticketsTable.userId, req.user!.userId))
        .orderBy(sql`${ticketsTable.createdAt} DESC`);
  res.json(rows.map(serialize));
});

router.post("/tickets", requireAuth, async (req, res) => {
  const body = CreateTicketBody.parse(req.body);
  const [t] = await db
    .insert(ticketsTable)
    .values({
      userId: req.user!.userId,
      subject: body.subject,
      description: body.description,
      priority: body.priority,
      relatedMessageId: body.relatedMessageId ?? null,
      externalId: `HEL-${Math.floor(Math.random() * 90000) + 10000}`,
      lastUpdate: "Ticket opened",
    })
    .returning();
  res.status(201).json(serialize(t));
});

router.get("/tickets/:id", requireAuth, async (req, res) => {
  const id = Number(req.params.id);
  const [t] = await db.select().from(ticketsTable).where(eq(ticketsTable.id, id));
  if (!t) {
    res.status(404).json({ error: "Ticket not found", status: 404 });
    return;
  }
  if (req.user!.role !== "admin" && t.userId !== req.user!.userId) {
    res.status(403).json({ error: "Forbidden", status: 403 });
    return;
  }
  res.json(serialize(t));
});

router.patch("/tickets/:id", requireAuth, async (req, res) => {
  const id = Number(req.params.id);
  const body = UpdateTicketBody.parse(req.body);
  const [existing] = await db.select().from(ticketsTable).where(eq(ticketsTable.id, id));
  if (!existing) {
    res.status(404).json({ error: "Ticket not found", status: 404 });
    return;
  }
  // Admin can change status; user can only update their own with last-message text.
  const isAdmin = req.user!.role === "admin";
  if (!isAdmin && existing.userId !== req.user!.userId) {
    res.status(403).json({ error: "Forbidden", status: 403 });
    return;
  }
  const updates: Partial<typeof ticketsTable.$inferInsert> = {};
  if (body.status !== undefined && isAdmin) updates.status = body.status;
  if (body.lastUpdate !== undefined) updates.lastUpdate = body.lastUpdate;
  if (Object.keys(updates).length === 0) {
    res.json(serialize(existing));
    return;
  }
  const [t] = await db.update(ticketsTable).set(updates).where(eq(ticketsTable.id, id)).returning();
  res.json(serialize(t));
});

export default router;

File: artifacts/api-server/app/routes/tickets.py
from __future__ import annotations

import logging
import random
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException

from app import zoho
from app.audit import audit_log
from app.auth import AuthedUser, require_auth
from app.db import get_db, next_id
from app.schemas import CreateTicketBody, UpdateTicketBody
from app.serialize import serialize_ticket

router = APIRouter()
log = logging.getLogger("api-server.tickets")


def _parse_id(raw: str) -> int:
    try:
        n = int(raw)
    except (TypeError, ValueError):
        raise HTTPException(status_code=400, detail="Invalid id")
    if n <= 0:
        raise HTTPException(status_code=400, detail="Invalid id")
    return n


@router.get("/tickets")
async def list_tickets(user: AuthedUser = Depends(require_auth)) -> list[dict[str, object]]:
    db = await get_db()
    is_admin = user.role == "admin"
    flt: dict[str, object] = {} if is_admin else {"userId": user.userId}
    rows = await db.tickets.find(flt).sort("createdAt", -1).to_list(length=None)
    return [serialize_ticket(r) for r in rows]


@router.get("/tickets/active-summary")
async def active_summary(user: AuthedUser = Depends(require_auth)) -> dict[str, object]:
    """Proactive open-ticket banner data for the chat sidebar.

    Refreshes Zoho-side status for the user's open tickets so the UI shows
    the latest update without the user having to navigate away.
    """
    db = await get_db()
    rows = await db.tickets.find(
        {"userId": user.userId, "status": {"$in": ["open", "in_progress"]}}
    ).sort("updatedAt", -1).to_list(length=10)

    if zoho.is_configured():
        for r in rows:
            ext = r.get("externalId")
            if not ext or not str(ext).startswith("zoho:"):
                continue
            zid = str(ext).split(":", 1)[1]
            data = await zoho.get_ticket(zid)
            if not data:
                continue
            new_status = (data.get("status") or "").lower().replace(" ", "_") or r["status"]
            new_update = data.get("statusType") or data.get("status") or r.get("lastUpdate")
            await db.tickets.update_one(
                {"_id": r["_id"]},
                {
                    "$set": {
                        "status": new_status,
                        "lastUpdate": new_update,
                        "updatedAt": datetime.now(timezone.utc),
                    }
                },
            )
            r["status"] = new_status
            r["lastUpdate"] = new_update

    summary = [
        {
            "id": r["_id"],
            "subject": r["subject"],
            "status": r["status"],
            "lastUpdate": r.get("lastUpdate"),
            "externalId": r.get("externalId"),
            "updatedAt": r.get("updatedAt"),
        }
        for r in rows
    ]
    return {"openCount": len(summary), "tickets": summary}


@router.post("/tickets", status_code=201)
async def create_ticket(
    body: CreateTicketBody, user: AuthedUser = Depends(require_auth)
) -> dict[str, object]:
    db = await get_db()
    now = datetime.now(timezone.utc)

    external_id = f"HEL-{random.randint(10000, 99999)}"
    if zoho.is_configured():
        try:
            resp = await zoho.create_ticket(
                subject=body.subject,
                description=body.description,
                priority=body.priority,
                requester_email=user.email,
                requester_name=" ".join(filter(None, [user.firstName, user.lastName])).strip() or user.email,
            )
            if resp and resp.get("id"):
                external_id = f"zoho:{resp['id']}"
        except Exception as err:
            log.warning("zoho create_ticket failed, keeping local id: %s", err)

    t = {
        "_id": await next_id("tickets"),
        "userId": user.userId,
        "subject": body.subject,
        "description": body.description,
        "priority": body.priority,
        "status": "open",
        "externalId": external_id,
        "relatedMessageId": body.relatedMessageId,
        "lastUpdate": "Ticket opened",
        "createdAt": now,
        "updatedAt": now,
    }
    await db.tickets.insert_one(t)
    await audit_log(
        action="ticket.create",
        actor=user.email or user.userId,
        target=external_id,
        meta={"priority": body.priority, "subject": body.subject},
    )
    return serialize_ticket(t)


@router.get("/tickets/{id}")
async def get_ticket(id: str, user: AuthedUser = Depends(require_auth)) -> dict[str, object]:
    tid = _parse_id(id)
    db = await get_db()
    t = await db.tickets.find_one({"_id": tid})
    if not t:
        raise HTTPException(status_code=404, detail="Ticket not found")
    if user.role != "admin" and t.get("userId") != user.userId:
        raise HTTPException(status_code=403, detail="Forbidden")
    return serialize_ticket(t)


@router.patch("/tickets/{id}")
async def update_ticket(
    id: str,
    body: UpdateTicketBody,
    user: AuthedUser = Depends(require_auth),
) -> dict[str, object]:
    tid = _parse_id(id)
    db = await get_db()
    existing = await db.tickets.find_one({"_id": tid})
    if not existing:
        raise HTTPException(status_code=404, detail="Ticket not found")
    is_admin = user.role == "admin"
    if not is_admin and existing.get("userId") != user.userId:
        raise HTTPException(status_code=403, detail="Forbidden")

    updates: dict[str, object] = {"updatedAt": datetime.now(timezone.utc)}
    if body.status is not None and is_admin:
        updates["status"] = body.status
    if body.lastUpdate is not None:
        updates["lastUpdate"] = body.lastUpdate

    r = await db.tickets.find_one_and_update(
        {"_id": tid}, {"$set": updates}, return_document=True
    )
    return serialize_ticket(r or existing)


@router.delete("/tickets/{id}", status_code=204)
async def delete_ticket(id: str, user: AuthedUser = Depends(require_auth)) -> None:
    tid = _parse_id(id)
    db = await get_db()
    existing = await db.tickets.find_one({"_id": tid})
    if not existing:
        raise HTTPException(status_code=404, detail="Ticket not found")

    is_admin = user.role == "admin"
    if not is_admin and existing.get("userId") != user.userId:
        raise HTTPException(status_code=403, detail="Forbidden")

    await db.tickets.delete_one({"_id": tid})
    await audit_log(
        action="ticket.delete",
        actor=user.email or user.userId,
        target=str(existing.get("externalId") or tid),
        meta={"ticketId": tid, "subject": existing.get("subject")},
    )
